### DISCUSSION TITLE
NewRoadTypes - level crossings and tunnels

### DISCUSSION THREAD

**Discussion by PikkaBird** [OpenTTD/OpenTTD#10237] [Features]
Some proposals coming out of #10223, #9931, etc, and fleshing out the NRT spec a little further.

Proposal 1: allow road and tramtypes to provide tunnel portal sprites via action 3. For a combined tunnel, the **tram** sprites are used.

> reldred: I dunno, I can imagine a tram tunnel perhaps wanting to be wider than a road tunnel, perhaps even slightly taller. And if I think of something like Wally’s suspended monorail that will really want the tram tunnel to override a road tunnel.

Proposal 2: allow road and tramtypes to provide level crossing sprites (overlay + gates) via action 3. If a road or tramtype provides these sprites, the gates are used in preference to the railtype's gates. For a combined road and tram crossing, the **road** gate sprites are used.

> Pruple: [...] generally there'll be a lot more different roadtypes, changing over years (dirt, gravel, paved, fancy, different markings over the years), each of which could conceivably have a different [...] boomgate style.

Proposal 3: expose information about a crossing tile's position in a multi-track crossing as a road and tramtype variable, to allow an appropriate road overlay to be drawn.

> #9931: When using a NewGRF railtype which includes custom crossing signal sprites, only the crossing signals at the edges of the crossing are drawn.

Thoughts?

**Comment by PikkaBird** [comment]
Proposal 4: add a property to road and tramtypes to specify an custom sound effect for the crossing bell. :) For a combined crossing, the road sound is used.

**Comment by kevinfields777** [comment]
Let me add in another thought. Why give any priority at all by default. Let the player decide by laying down the sprites with the most recent action. If they don't like how it looks, they can flip back over to the type they prefer and overlay that instead.

**Reply by PikkaBird** [reply]
I'm not sure I'm a fan of that idea. Newgrf authors will always have a "correct" crossing or tunnel in mind for any given combination, and requiring players to build things in a particular order to get it is both clunky and non-obvious from a UX perspective.

Apart from that, it'd add the additional complexity of requiring the tile to remember which transport type was built last.

**Comment by LC-Zorg** [comment]
I think forcing change of road tunnel graphics when a player adds tram tracks to it can be unexpected and even annoying. The more so that it also seems unlikely that the graphics of the road tunnel would not match the graphics of the tram tracks. Suspended Monorail is currently probably the only case. I think it would be better if the graphical change only happened as a last resort, when it is necessary to combine selected types of roads, or preferably if the graphical change was always the sole result of the player's decision.

Another solution could be that when building tunnels, as in the case of bridges, player can choose the type of tunnel. Then each tunnel graphic could be assigned (beyond cost and speed limit) types of compatible roads, tracks, wires or whatever else you come up with. If a certain type of tunnel was not intended for the selected road type, it might be greyed out or simply not visible in the selection window.
![Tunnels 3](https://user-images.githubusercontent.com/67662046/207891923-e4e51950-3a8c-4f60-8aaa-5009d537c36e.png)
or
![Tunnels 4](https://user-images.githubusercontent.com/67662046/207891992-75a68107-abe7-48a0-a4aa-81c2109dd558.png)

**Reply by PikkaBird** [reply]
I think the objective here should be to add to NRT those NewGRF features which already exist for railtypes, in as simple and lightweight a way as possible. If you want to create a patch for tunneltypes, that's an entirely separate thing.

As you say, I suspect most newgrf authors would _not_ use different tunnel graphics for tram and road tunnels, but - in cases where they do - one has to take precedence. My original thought, as with level crossings, was that it should be roads, because they'll have more variety, but reldred suggested it should be tram.

As a suggestion, how about reverting to roadtype takes precedence, unless the tramtype has a flag set in prop 10 to say "always use my tunnel"? That allows for either to take precedence based on the particular design of the newgrf set. A similar flag could be used for level crossings.
